File: src/feed.ts
import Parser from "rss-parser";
import type { FeedItem } from "./types.js";

/**
 * rss-parser custom field output used by this application.
 */
interface ParsedItem {
	/**
	 * Feed GUID.
	 */
	guid?: string;
	/**
	 * Item title.
	 */
	title?: string;
	/**
	 * Item URL.
	 */
	link?: string;
	/**
	 * RSS content snippet or Atom summary.
	 */
	contentSnippet?: string;
	/**
	 * RSS content HTML.
	 */
	content?: string;
	/**
	 * RSS description HTML.
	 */
	description?: string;
	/**
	 * Publication date string.
	 */
	pubDate?: string;
	/**
	 * ISO publication date string added by rss-parser.
	 */
	isoDate?: string;
	/**
	 * Atom updated timestamp.
	 */
	updated?: string;
	/**
	 * Creator field from RSS modules.
	 */
	creator?: string;
	/**
	 * Author field.
	 */
	author?: string;
}

/**
 * Shared RSS and Atom parser.
 */
const parser = new Parser<unknown, ParsedItem>({
	headers: {
		"User-Agent": "rss-discord/1.0",
		Accept:
			"application/rss+xml, application/atom+xml, application/xml, text/xml",
	},
	timeout: 15000,
});

/**
 * Fetches a feed URL and normalizes all entries that have an identifier.
 */
export async function fetchFeed(url: string): Promise<FeedItem[]> {
	let parsed: Parser.Output<ParsedItem>;
	try {
		parsed = await parser.parseURL(url);
	} catch (error) {
		throw new Error(`フィード取得失敗: ${formatError(error)}`);
	}

	return parsed.items
		.map((item) => normalizeItem(item))
		.filter((item) => item.id !== "");
}

/**
 * Normalizes one parser item into the app's stable feed item shape.
 */
function normalizeItem(item: ParsedItem): FeedItem {
	const id = item.guid ?? item.link ?? item.title ?? "";
	const title = item.title && item.title !== "" ? item.title : "Untitled";
	const pubDate = normalizeDate(
		item.isoDate ?? item.pubDate ?? item.updated ?? "",
	);

	return {
		id,
		title,
		link: item.link ?? "",
		description: item.contentSnippet ?? item.description ?? item.content ?? "",
		pubDate,
		author: item.creator ?? item.author ?? "",
	};
}

/**
 * Converts parseable dates to ISO 8601 while preserving unparseable input.
 */
function normalizeDate(value: string): string {
	if (value === "") {
		return "";
	}

	const timestamp = Date.parse(value);
	if (Number.isNaN(timestamp)) {
		return value;
	}
	return new Date(timestamp).toISOString();
}

/**
 * Formats unknown parser errors.
 */
function formatError(error: unknown): string {
	return error instanceof Error ? error.message : String(error);
}

File: internal/feed/feed.go
package feed

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/mmcdole/gofeed"
)

type FeedItem struct {
	ID          string
	Title       string
	Link        string
	Description string
	PubDate     string
	Author      string
}

func FetchFeed(ctx context.Context, url string) ([]FeedItem, error) {
	ctx, cancel := context.WithTimeout(ctx, 15*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("リクエスト作成失敗: %w", err)
	}
	req.Header.Set("User-Agent", "rss-discord/1.0")
	req.Header.Set("Accept", "application/rss+xml, application/atom+xml, application/xml, text/xml")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("フィード取得失敗: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		_, _ = io.ReadAll(resp.Body)
		return nil, fmt.Errorf("HTTP %d fetching %s", resp.StatusCode, url)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("レスポンス読み取り失敗: %w", err)
	}

	parser := gofeed.NewParser()
	parsed, err := parser.ParseString(string(body))
	if err != nil {
		return nil, fmt.Errorf("フィードパース失敗: %w", err)
	}

	items := make([]FeedItem, 0, len(parsed.Items))
	for _, item := range parsed.Items {
		fi := normalize(item)
		if fi.ID != "" {
			items = append(items, fi)
		}
	}

	return items, nil
}

func normalize(item *gofeed.Item) FeedItem {
	id := item.GUID
	if id == "" {
		id = item.Link
	}
	if id == "" {
		id = item.Title
	}

	title := item.Title
	if title == "" {
		title = "Untitled"
	}

	var pubDate string
	if item.PublishedParsed != nil {
		pubDate = item.PublishedParsed.Format(time.RFC3339)
	} else if item.UpdatedParsed != nil {
		pubDate = item.UpdatedParsed.Format(time.RFC3339)
	} else if item.Published != "" {
		pubDate = item.Published
	}

	var author string
	if item.Author != nil {
		author = item.Author.Name
	}

	return FeedItem{
		ID:          id,
		Title:       title,
		Link:        item.Link,
		Description: item.Description,
		PubDate:     pubDate,
		Author:      author,
	}
}
